File: datamodelgov-examples/go-client-example/real_thrift_example.go
package main

import (
	"fmt"
	"time"
	"context"
	"git.apache.org/thrift.git/lib/go/thrift"
	"github.com/tsinghua/datamodelgov-sdk-go/src/tsinghua/api"
)

func main() {
	fmt.Println("=== DataModelGov Go Thrift SDK 客户端示例 ===\n")
	
	// 1. 创建Thrift连接
	fmt.Println("1. 连接到DataModelGov服务端")
	transport, err := thrift.NewTSocket("localhost:9090")
	if err != nil {
		fmt.Printf("创建传输失败: %v\n", err)
		return
	}
	defer transport.Close()
	
	protocolFactory := thrift.NewTBinaryProtocolFactoryDefault()
	client := api.NewApiServiceClientFactory(transport, protocolFactory)
	
	// 打开连接
	if err := transport.Open(); err != nil {
		fmt.Printf("连接失败: %v\n", err)
		return
	}
	fmt.Println("✅ 连接成功: localhost:9090")
	fmt.Println()
	
	// 2. 创建数据模型元数据
	fmt.Println("2. 创建数据模型元数据")
	modelMeta := api.NewModelMeta()
	modelMeta.Name = "用户模型"
	modelMeta.Version = "1.0.0"
	modelMeta.Author = "data_scientist_001"
	modelMeta.Scene = "用户行为分析"
	modelMeta.Inputs = "{\"age\": \"int\", \"income\": \"double\"}"
	modelMeta.Outputs = "{\"risk_score\": \"double\", \"category\": \"string\"}"
	modelMeta.Timestamp = time.Now().Unix()
	
	fmt.Printf("模型名称: %s\n", modelMeta.Name)
	fmt.Printf("模型版本: %s\n", modelMeta.Version)
	fmt.Printf("模型作者: %s\n", modelMeta.Author)
	fmt.Println()
	
	// 3. 调用保存模型元数据API
	fmt.Println("3. 调用saveModelMeta API")
	ctx := context.Background()
	saveResult, err := client.SaveModelMeta(ctx, modelMeta)
	if err != nil {
		fmt.Printf("保存失败: %v\n", err)
	} else {
		fmt.Printf("保存结果: Success=%t, Message=%s\n", saveResult.Success, saveResult.Message)
	}
	fmt.Println()
	
	// 4. 调用查询模型元数据API
	fmt.Println("4. 调用getModelMeta API")
	retrievedResult, err := client.GetModelMeta(ctx, "用户模型", "1.0.0")
	if err != nil {
		fmt.Printf("查询失败: %v\n", err)
	} else {
		fmt.Printf("查询结果: Success=%t, Message=%s\n", retrievedResult.Success, retrievedResult.Message)
		if retrievedResult.Data != nil {
			fmt.Printf("数据: %s\n", *retrievedResult.Data)
		}
	}
	fmt.Println()
	
	// 5. 创建数据查询请求
	fmt.Println("5. 创建数据查询请求")
	queryRequest := api.NewDataQueryRequest()
	queryRequest.Paths = []string{"root.device.temperature", "root.device.humidity"}
	queryRequest.StartTime = time.Now().Unix() - 86400 // 24小时前
	queryRequest.EndTime = time.Now().Unix()
	
	fmt.Printf("查询路径: %v\n", queryRequest.Paths)
	fmt.Printf("时间范围: %d - %d\n", queryRequest.StartTime, queryRequest.EndTime)
	fmt.Println()
	
	// 6. 调用数据查询API
	fmt.Println("6. 调用queryData API")
	queryResult, err := client.QueryData(ctx, queryRequest)
	if err != nil {
		fmt.Printf("查询失败: %v\n", err)
	} else {
		fmt.Printf("查询结果: Success=%t, Message=%s\n", queryResult.Success, queryResult.Message)
		if queryResult.Data != nil {
			fmt.Printf("数据: %s\n", *queryResult.Data)
		}
	}
	fmt.Println()
	
	// 7. 演示其他数据对象
	fmt.Println("7. 演示其他数据对象")
	
	// 创建关联规则查询请求
	rulesQuery := api.NewAssociationRulesQueryRequest()
	rulesQuery.PageNum = 1
	rulesQuery.PageSize = 10
	rulesQuery.Name = "用户行为规则"
	
	// 创建存储引擎信息
	storageInfo := api.NewStorageEngineInfo()
	storageInfo.Id = 1
	storageInfo.Ip = "127.0.0.1"
	storageInfo.Port = 8080
	storageInfo.Type = 1 // IoTDB
	storageInfo.SchemaPrefix = "root"
	storageInfo.DataPrefix = "data"
	
	fmt.Printf("关联规则查询: PageNum=%d, PageSize=%d\n", rulesQuery.PageNum, rulesQuery.PageSize)
	fmt.Printf("存储引擎信息: IP=%s:%d, Type=%d\n", storageInfo.Ip, storageInfo.Port, storageInfo.Type)
	fmt.Println()
	
	// 8. 演示字段枚举
	fmt.Println("8. SDK字段枚举")
	fmt.Println("ModelMeta字段:")
	for _, field := range api.ModelMetaFields {
		fmt.Printf("  - %s (thriftId: %d)\n", field.String(), int16(field))
	}
	fmt.Println()
	
	fmt.Println("✅ 所有Go SDK调用演示完成")
	fmt.Println("=== 示例完成 ===")
}

File: datamodelgov-examples/python-client-example/real_thrift_example.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
DataModelGov Python Thrift SDK 客户端示例
使用真正的Thrift SDK进行API调用
"""

import sys
import os
import time
import logging
from thrift.transport import TSocket
from thrift.transport import TTransport
from thrift.protocol import TBinaryProtocol

# 添加SDK路径
sys.path.append(os.path.join(os.path.dirname(__file__), '../../datamodelgov-sdk-python/src'))

try:
    from tsinghua.api import ApiService
    from tsinghua.api.ttypes import ModelMeta, DataQueryRequest, AssociationRulesQueryRequest, StorageEngineInfo
except ImportError as e:
    print(f"导入SDK失败: {e}")
    print("请确保Python SDK已正确安装")
    sys.exit(1)

# 配置日志
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)

def main():
    print("=== DataModelGov Python Thrift SDK 客户端示例 ===\n")
    
    # 1. 创建Thrift连接
    print("1. 连接到DataModelGov服务端")
    try:
        transport = TSocket.TSocket('localhost', 9090)
        protocol = TBinaryProtocol.TBinaryProtocol(transport)
        client = ApiService.Client(protocol)
        transport.open()
        print("✅ 连接成功: localhost:9090")
        print()
    except Exception as e:
        print(f"❌ 连接失败: {e}")
        return
    
    try:
        # 2. 创建数据模型元数据
        print("2. 创建数据模型元数据")
        model_meta = ModelMeta()
        model_meta.name = "用户模型"
        model_meta.version = "1.0.0"
        model_meta.author = "data_scientist_001"
        model_meta.scene = "用户行为分析"
        model_meta.inputs = '{"age": "int", "income": "double"}'
        model_meta.outputs = '{"risk_score": "double", "category": "string"}'
        model_meta.timestamp = int(time.time())
        
        print(f"模型名称: {model_meta.name}")
        print(f"模型版本: {model_meta.version}")
        print(f"模型作者: {model_meta.author}")
        print()
        
        # 3. 调用保存模型元数据API
        print("3. 调用saveModelMeta API")
        try:
            save_result = client.saveModelMeta(model_meta)
            print(f"保存结果: Success={save_result.success}, Message={save_result.message}")
        except Exception as e:
            print(f"保存失败: {e}")
        print()
        
        # 4. 调用查询模型元数据API
        print("4. 调用getModelMeta API")
        try:
            retrieved_result = client.getModelMeta("用户模型", "1.0.0")
            print(f"查询结果: Success={retrieved_result.success}, Message={retrieved_result.message}")
            if retrieved_result.data:
                print(f"数据: {retrieved_result.data}")
        except Exception as e:
            print(f"查询失败: {e}")
        print()
        
        # 5. 创建数据查询请求
        print("5. 创建数据查询请求")
        query_request = DataQueryRequest()
        query_request.paths = ["root.device.temperature", "root.device.humidity"]
        query_request.start_time = int(time.time()) - 86400  # 24小时前
        query_request.end_time = int(time.time())
        
        print(f"查询路径: {query_request.paths}")
        print(f"时间范围: {query_request.start_time} - {query_request.end_time}")
        print()
        
        # 6. 调用数据查询API
        print("6. 调用queryData API")
        try:
            query_result = client.queryData(query_request)
            print(f"查询结果: Success={query_result.success}, Message={query_result.message}")
            if query_result.data:
                print(f"数据: {query_result.data}")
        except Exception as e:
            print(f"查询失败: {e}")
        print()
        
        # 7. 演示其他数据对象
        print("7. 演示其他数据对象")
        
        # 创建关联规则查询请求
        rules_query = AssociationRulesQueryRequest()
        rules_query.page_num = 1
        rules_query.page_size = 10
        rules_query.name = "用户行为规则"
        
        # 创建存储引擎信息
        storage_info = StorageEngineInfo()
        storage_info.id = 1
        storage_info.ip = "127.0.0.1"
        storage_info.port = 8080
        storage_info.type = 1  # IoTDB
        storage_info.schema_prefix = "root"
        storage_info.data_prefix = "data"
        
        print(f"关联规则查询: PageNum={rules_query.page_num}, PageSize={rules_query.page_size}")
        print(f"存储引擎信息: IP={storage_info.ip}:{storage_info.port}, Type={storage_info.type}")
        print()
        
        # 8. 演示字段枚举
        print("8. SDK字段枚举")
        print("ModelMeta字段:")
        if hasattr(ModelMeta, 'THRIFT_SPEC'):
            for field_id in ModelMeta.THRIFT_SPEC:
                print(f"  - Field ID: {field_id}")
        print()
        
        print("✅ 所有Python SDK调用演示完成")
        
    except Exception as e:
        print(f"❌ 操作失败: {e}")
        logger.exception("详细错误信息")
    
    finally:
        # 9. 关闭连接
        if transport.isOpen():
            transport.close()
            print("✅ 连接已关闭")
    
    print("=== 示例完成 ===")

if __name__ == "__main__":
    main()
